ext/imap/php_imap.c

```diff
@@ -695,18 +700,29 @@ PHP_MINFO_FUNCTION(imap)
 }
 /* }}} */
 
-/* {{{ imap_do_open */
-static void php_imap_do_open(INTERNAL_FUNCTION_PARAMETERS, int persistent)
+/* {{{ Open an IMAP stream to a mailbox */
+PHP_FUNCTION(imap_open)
 {
 	zend_string *mailbox, *user, *passwd;
 	zend_long retries = 0, flags = NIL, cl_flags = NIL;
 	MAILSTREAM *imap_stream;
 	pils *imap_le_struct;
-	zval *params = NULL;
+	HashTable *params = NULL;
 	int argc = ZEND_NUM_ARGS();
 
-	if (zend_parse_parameters(argc, "PSS|lla", &mailbox, &user,
-		&passwd, &flags, &retries, &params) == FAILURE) {
+	if (zend_parse_parameters(argc, "PSS|llh", &mailbox, &user, &passwd, &flags, &retries, &params) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	if (flags && ((flags & ~(OP_READONLY | OP_ANONYMOUS | OP_HALFOPEN | CL_EXPUNGE | OP_DEBUG | OP_SHORTCACHE
+	 		| OP_SILENT | OP_PROTOTYPE | OP_SECURE)) != 0)) {
+		zend_argument_value_error(4, "must be a bitmask of OP_READONLY, OP_ANONYMOUS, OP_HALFOPEN, CL_EXPUNGE, "
+			"OP_DEBUG, OP_SHORTCACHE, OP_SILENT, OP_PROTOTYPE, and OP_SECURE");
```

Comment:
```suggestion
		zend_argument_value_error(4, "must be a bitmask of OP_* constants");
```
I believe we don't list all options explicitly if there are so many. @kocsismate Is the format right?

Reply:
Yes, currently we make use of `*` if there are more than 3 possible values, unless using `*` is not possible.